runc tests fail with criu-dev

For the last few years, runc repo have a CI job to test runc against criu-dev branch, mostly to catch some regressions early.

We are seeing failures recently (this week), alas without too much diagnostics.

For example (from [here](https://productionresultssa6.blob.core.windows.net/actions-results/865121f5-ae53-4a5d-a72a-1f3aace783d2/workflow-job-run-15406e29-e965-5079-8b70-bbc24930c4cb/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-10-16T17%3A41%3A00Z&sig=F9TpGdybzQNGVNqU1oT0IKHCEtLAy%2BqSiUzOFRMHqt8%3D&ske=2025-10-17T05%3A10%3A53Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-10-16T17%3A10%3A53Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-11-05&sp=r&spr=https&sr=b&st=2025-10-16T17%3A30%3A55Z&sv=2025-11-05), look for `not ok 32`):
```
2025-10-16T01:03:57.5662977Z not ok 32 checkpoint and restore
2025-10-16T01:03:57.5673614Z # (from function `simple_cr' in file tests/integration/checkpoint.bats, line 131,
2025-10-16T01:03:57.5675071Z #  in test file tests/integration/checkpoint.bats, line 192)
2025-10-16T01:03:57.5698743Z #   `simple_cr' failed
2025-10-16T01:03:57.5712389Z # runc spec (status=0)
2025-10-16T01:03:57.5713588Z #
2025-10-16T01:03:57.5714281Z # runc run -d --console-socket /tmp/bats-run-kiFEqC/runc.CtdpiC/tty/sock test_busybox (status=0)
2025-10-16T01:03:57.5715346Z #
2025-10-16T01:03:57.5715677Z # runc state test_busybox (status=0)
2025-10-16T01:03:57.5716304Z # {
2025-10-16T01:03:57.5716613Z #   "ociVersion": "1.2.1+dev",
2025-10-16T01:03:57.5716999Z #   "id": "test_busybox",
2025-10-16T01:03:57.5717360Z #   "pid": 13992,
2025-10-16T01:03:57.5717695Z #   "status": "running",
2025-10-16T01:03:57.5718167Z #   "bundle": "/tmp/bats-run-kiFEqC/runc.CtdpiC/bundle",
2025-10-16T01:03:57.5718804Z #   "rootfs": "/tmp/bats-run-kiFEqC/runc.CtdpiC/bundle/rootfs",
2025-10-16T01:03:57.5719398Z #   "created": "2025-10-16T01:03:57.172483254Z",
2025-10-16T01:03:57.5719845Z #   "owner": ""
2025-10-16T01:03:57.5720295Z # }
2025-10-16T01:03:57.5720817Z # runc checkpoint --work-path ./work-dir test_busybox (status=1)
2025-10-16T01:03:57.5721697Z # time="2025-10-16T01:03:57Z" level=warning msg="open work-dir/dump.log: no such file or directory"
2025-10-16T01:03:57.5722327Z # time="2025-10-16T01:03:57Z" level=error msg="read unixpacket @->@: EOF"
2025-10-16T01:03:57.5722676Z # --- teardown ---
```

It seems that criu swrk just dies under (without even leaving `dump.log`).

Tried to reproduce it locally with no luck so far.

> This fairly innocuous PR (https://github.com/opencontainers/runc/pull/4973) appears to make the problem worse somehow (unless I'm just really unlucky, this job failed >10 times in a row when usually re-running the job once or twice is enough to get it to be green).

@cyphar This error looks different:

```
2025-11-08T03:20:03.7700585Z     exec_test.go:1748: extra fd 9 -> /9572/task/10780/fd
2025-11-08T03:20:03.7701338Z     exec_test.go:1751: found 1 extra fds after container.Run
2025-11-08T03:20:03.7702097Z --- FAIL: TestFdLeaksSystemd (0.09s)
```

--- a/criu/log.c
+++ b/criu/log.c
@@ -190,7 +190,7 @@ void flush_early_log_buffer(int fd)
 		 * with reading the log_level.
 		 */
 		struct early_log_hdr *hdr = (void *)early_log_buffer + pos;
-		pos += sizeof(hdr);
+		pos += sizeof(*hdr);
 		if (hdr->level <= current_loglevel) {
 			size_t size = 0;
 			while (size < hdr->len) {
@@ -323,7 +323,7 @@ static void early_vprint(const char *format, unsigned int loglevel, va_list para
 	int log_size = 0, log_space;
 	struct early_log_hdr *hdr;
 
-	if ((early_log_buf_off + sizeof(hdr)) >= EARLY_LOG_BUF_LEN)
+	if ((early_log_buf_off + sizeof(*hdr)) >= EARLY_LOG_BUF_LEN)
 		return;
 
 	/* Save loglevel */
@@ -331,7 +331,7 @@ static void early_vprint(const char *format, unsigned int loglevel, va_list para
 	hdr = (void *)early_log_buffer + early_log_buf_off;
 	hdr->level = loglevel;
 	/* Skip the log entry size */
-	early_log_buf_off += sizeof(hdr);
+	early_log_buf_off += sizeof(*hdr);
 	log_space = EARLY_LOG_BUF_LEN - early_log_buf_off;
 	if (loglevel >= LOG_TIMESTAMP) {
 		/*

--- a/criu/log.c
+++ b/criu/log.c
@@ -202,7 +202,7 @@ void flush_early_log_buffer(int fd)
 		}
 		pos += hdr->len;
 	}
-	if (early_log_buf_off == EARLY_LOG_BUF_LEN)
+	if ((early_log_buf_off + sizeof(struct early_log_hdr)) >= EARLY_LOG_BUF_LEN)
 		pr_warn("The early log buffer is full, some messages may have been lost\n");
 	early_log_buf_off = 0;
 }
@@ -320,7 +320,7 @@ unsigned int log_get_loglevel(void)
 
 static void early_vprint(const char *format, unsigned int loglevel, va_list params)
 {
-	unsigned int log_size = 0;
+	int log_size = 0, log_space;
 	struct early_log_hdr *hdr;
 
 	if ((early_log_buf_off + sizeof(hdr)) >= EARLY_LOG_BUF_LEN)
@@ -332,6 +332,7 @@ static void early_vprint(const char *format, unsigned int loglevel, va_list para
 	hdr->level = loglevel;
 	/* Skip the log entry size */
 	early_log_buf_off += sizeof(hdr);
+	log_space = EARLY_LOG_BUF_LEN - early_log_buf_off;
 	if (loglevel >= LOG_TIMESTAMP) {
 		/*
 		 * If logging is not yet setup we just write zeros
@@ -339,12 +340,17 @@ static void early_vprint(const char *format, unsigned int loglevel, va_list para
 		 * keep the same format as the other messages on
 		 * log levels with timestamps (>=LOG_TIMESTAMP).
 		 */
-		log_size = snprintf(early_log_buffer + early_log_buf_off, sizeof(early_log_buffer) - early_log_buf_off,
+		log_size = snprintf(early_log_buffer + early_log_buf_off, log_space,
 				    "(00.000000) ");
 	}
 
-	log_size += vsnprintf(early_log_buffer + early_log_buf_off + log_size,
-			      sizeof(early_log_buffer) - early_log_buf_off - log_size, format, params);
+	if (log_size < log_space)
+		log_size += vsnprintf(early_log_buffer + early_log_buf_off + log_size,
+				      log_space - log_size, format, params);
+	if (log_size > log_space) {
+		/* vsnprintf always add the terminating null byte. */
+		log_size = log_space - 1;
+	}
 
 	/* Save log entry size */
 	hdr->len = log_size;

